Removing ArrayTypeDesc

VSadov opened the pull request:
At this point ArrayTypeDesc is an unnecessary indirection.

VSadov committed:
Reduce use of ArrayTpeDesc, somewhat

VSadov committed:
remove `AsArray`. Fewer uses of ArrayTypeDesc

VSadov committed:
remove `IsArrayType`

VSadov committed:
remove AssertArrayTypeDescLoaded

VSadov committed:
normalized use of GetElementType vs GetTypeParam vs GetArrayElementTypeHandle

VSadov committed:
removed GetMethodTableOfElementType

VSadov committed:
HasTypeParam  and GetTypeParam should work for TypeHandle when array is not a TypeDesc

VSadov committed:
Some cleanup of `IsTypeDesc` use.

VSadov committed:
Removed `ArrayTypeDesc`

VSadov committed:
a few fixes

VSadov committed:
another fix

VSadov commented:
The last fix was pretty darn hard to debug.

VSadov committed:
dealt with TODOs

VSadov commented:
I think this PR is ready for review.

VSadov commented on src/coreclr/src/vm/clsload.cpp:
```diff
@@ -4935,7 +4930,7 @@ BOOL ClassLoader::CanAccessMethodInstantiation( // True if access is legal, fals
             TypeHandle th = inst[i];
 
-            MethodTable* pMT = th.GetMethodTableOfElementType();
+            MethodTable* pMT = th.GetMethodTableOfRootElementType();
```
`GetMethodTableOfElementType` was renamed to match closer what it does.

VSadov commented on src/coreclr/src/vm/gdbjit.cpp:
```diff
@@ -152,7 +152,7 @@ GetTypeInfoFromTypeHandle(TypeHandle typeHandle,
         case ELEMENT_TYPE_BYREF:
         {
-            TypeInfoBase* valTypeInfo = GetTypeInfoFromTypeHandle(typeHandle.GetTypeParam(), pTypeMap, method);
+            TypeInfoBase* valTypeInfo = GetTypeInfoFromTypeHandle(typeHandle.GetArrayElementTypeHandle(), pTypeMap, method);
```
should be `GetTypeParam`

VSadov commented on src/coreclr/src/vm/generics.cpp:
```diff
@@ -800,10 +800,9 @@ void RecursionGraph::AddDependency(MethodTable *pMT, TypeHandleList *pExpansionV
             {
-                _ASSERTE(thArg.HasTypeParam());
-                thArg = (static_cast<PTR_ParamTypeDesc>(thArg.AsTypeDesc()))->GetModifiedType();
+                thArg = thArg.GetArrayElementTypeHandle();
```
should be `GetTypeParam`

VSadov committed:
couple fixes from reviewing code

VSadov commented:
/azp run

jkotas commented on src/coreclr/src/vm/i386/stublinkerx86.cpp:
```diff
@@ -4491,26 +4491,18 @@ VOID StubLinkerCPU::EmitArrayOpStub(const ArrayOpScript* pArrayOpScript)
             X86EmitOp(0x8b, kArrayMTReg, kArrayRefReg, 0 AMD64_ARG(k64BitOp));
-            // cmp typeReg, [ESI/R10+m_ElementType];
-            X86EmitOp(0x3b, typeReg, kArrayMTReg, MethodTable::GetOffsetOfArrayElementTypeHandle() AMD64_ARG(k64BitOp));
+            // cmp typeReg, kArrayMTReg
```
Nit: I think this can be folded into previous instruction

VSadov replied:
I thought this could be `cmp typeReg, [kArrayRefReg]`, but was not sure. The API seems to require good knowledge of x86 encoding (to the level of `0x3b` being `cmp r, r` ), so I opted for copying something from few lines above.
I will try folding.

jkotas commented on src/coreclr/src/vm/ceeload.cpp:
```diff
@@ -3253,7 +3253,7 @@ void Module::FreeClassTables()
                 else if (th.IsArray())
                 {
-                    ComCallWrapperTemplate *pTemplate = th.AsArray()->GetComCallWrapperTemplate();
+                    ComCallWrapperTemplate *pTemplate = th.GetMethodTable()->GetComCallWrapperTemplate();
```
```suggestion
                    ComCallWrapperTemplate *pTemplate = th.AsMethodTable()->GetComCallWrapperTemplate();
```

jkotas commented on src/coreclr/src/vm/ceeload.cpp:
```diff
@@ -3253,7 +3253,7 @@ void Module::FreeClassTables()
                 }
                 else if (th.IsArray())
```
This path should be unreachable now. Delete it?

jkotas commented on src/coreclr/src/vm/gcenv.ee.cpp:
```diff
@@ -1425,7 +1425,7 @@ void GCToEEInterface::WalkAsyncPinnedForPromotion(Object* object, ScanContext* s
     if (pOverlapped->m_userObject != NULL)
     {
-        if (pOverlapped->m_userObject->GetGCSafeMethodTable() == g_pPredefinedArrayTypes[ELEMENT_TYPE_OBJECT]->GetMethodTable())
+        if (pOverlapped->m_userObject->GetGCSafeMethodTable() == g_pPredefinedArrayTypes[ELEMENT_TYPE_OBJECT].GetMethodTable())
```
```suggestion
        if (pOverlapped->m_userObject->GetGCSafeMethodTable() == g_pPredefinedArrayTypes[ELEMENT_TYPE_OBJECT].AsMethodTable())
```

jkotas commented on src/coreclr/src/vm/gcenv.ee.cpp:
```diff
@@ -1464,7 +1464,7 @@ void GCToEEInterface::WalkAsyncPinned(Object* object, void* context, void (*call
         Object * pUserObject = OBJECTREFToObject(pOverlapped->m_userObject);
         callback(object, pUserObject, context);
-        if (pOverlapped->m_userObject->GetGCSafeMethodTable() == g_pPredefinedArrayTypes[ELEMENT_TYPE_OBJECT]->GetMethodTable())
+        if (pOverlapped->m_userObject->GetGCSafeMethodTable() == g_pPredefinedArrayTypes[ELEMENT_TYPE_OBJECT].GetMethodTable())
```
```suggestion
        if (pOverlapped->m_userObject->GetGCSafeMethodTable() == g_pPredefinedArrayTypes[ELEMENT_TYPE_OBJECT].AsMethodTable())
```

jkotas commented on src/coreclr/src/vm/mlinfo.cpp:
```diff
@@ -2944,7 +2940,7 @@ MarshalInfo::MarshalInfo(Module* pModule,
             m_args.na.m_pArrayMT = arrayTypeHnd.GetMethodTable();
```
```suggestion
            m_args.na.m_pArrayMT = arrayTypeHnd.AsMethodTable();
```

jkotas commented on src/coreclr/src/vm/runtimehandles.cpp:
```diff
@@ -448,8 +448,7 @@ FCIMPL1(MethodDesc *, RuntimeTypeHandle::GetFirstIntroducedMethod, ReflectClassB
+        return NULL;
     }
 
     MethodTable* pMT = typeHandle.GetMethodTable();
```
```suggestion
    MethodTable* pMT = typeHandle.AsMethodTable();
```

jkotas commented on src/coreclr/src/vm/runtimehandles.cpp:
```diff
@@ -1050,7 +1043,7 @@ FCIMPL1(LPCUTF8, RuntimeTypeHandle::GetUtf8Name, ReflectClassBaseObject* pTypeUN
+    if (typeHandle.IsTypeDesc() || typeHandle.IsArray())
         FCThrowRes(kArgumentException, W("Arg_InvalidHandle"));
 
     MethodTable* pMT= typeHandle.GetMethodTable();
```
```suggestion
    MethodTable* pMT= typeHandle.AsMethodTable();
```

jkotas commented on src/coreclr/src/vm/typehandle.inl:
```diff
@@ -38,48 +38,33 @@ inline void TypeHandle::SetIsFullyLoaded()
-        return AsMethodTable();
-    }
+    TypeHandle current = *this;
+    while (current.HasTypeParam())
```
What about array of pointers? This method will return the type that the pointer points to - not an element type.

VSadov replied:
I think this is intentional. The helper is used for accessibility checks and such. It basically digs through constructed types to the concrete type parameter to delegate the checks. 
It should unwrap array of pointers to the pointed type.

jkotas replied:
Yes, I agree. The method should be called ...RootTypeParam... then

jkotas commented on src/coreclr/src/vm/typedesc.cpp:
```diff
@@ -74,10 +57,10 @@ TypeHandle TypeDesc::GetBaseTypeParam()
 
```
This is same as `GetMethodTableOfRootElementType`. It would be nice if it is called same (`Root` or `Base`) in both cases.

VSadov replied:
I like `Root` more. - A better hint that it is recursive (vs. just one layer). Will rename this one.

jkotas commented on src/coreclr/src/debug/daccess/request.cpp:
```diff
@@ -3119,9 +3119,9 @@ ClrDataAccess::GetUsefulGlobals(struct DacpUsefulGlobalsData *globalsData)
-        globalsData->ArrayMethodTable = HOST_CDADDR(objArray->GetMethodTable());
+    TypeHandle objArray = g_pPredefinedArrayTypes[ELEMENT_TYPE_OBJECT];
+    if (objArray != NULL)
+        globalsData->ArrayMethodTable = HOST_CDADDR(objArray.GetMethodTable());
```
```suggestion
        globalsData->ArrayMethodTable = HOST_CDADDR(objArray.AsMethodTable());
```

jkotas commented on src/coreclr/src/vm/gchelpers.cpp:
```diff
@@ -415,8 +415,7 @@ OBJECTREF AllocateSzArray(TypeHandle arrayType, INT32 cElements, GC_ALLOC_FLAGS
 
-    ArrayTypeDesc* arrayDesc = arrayType.AsArray();
-    MethodTable* pArrayMT = arrayDesc->GetMethodTable();
+    MethodTable* pArrayMT = arrayType.GetMethodTable();
```
```suggestion
    MethodTable* pArrayMT = arrayType.AsMethodTable();
```

jkotas commented on src/coreclr/src/vm/gchelpers.cpp:
```diff
@@ -615,8 +614,7 @@ OBJECTREF AllocateArrayEx(TypeHandle arrayType, INT32 *pArgs, DWORD dwNumArgs, G
 
-    ArrayTypeDesc* arrayDesc = arrayType.AsArray();
-    MethodTable* pArrayMT = arrayDesc->GetMethodTable();
+    MethodTable* pArrayMT = arrayType.GetMethodTable();
```
```suggestion
    MethodTable* pArrayMT = arrayType.AsMethodTable();
```

jkotas commented on src/coreclr/src/vm/object.inl:
```diff
@@ -183,7 +183,7 @@ inline /* static */ TypeHandle ArrayBase::GetTypeHandle(MethodTable * pMT)
     CorElementType kind = pMT->GetInternalCorElementType();
```
Can this just return the pMT now? It should not need to go through the type loader.

VSadov replied:
Indeed, can just return MT.

VSadov replied:
In fact the whole `ArrayBase::GetTypeHandle` deal is unnecessary. Same implementation as `Object`

VSadov replied:
`Object::GetTypeHandle()` is much simpler now. `GetTrueTypeHandle` is gone. `ArrayBase::GetTypeHandle` is gone too.

VSadov commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -3848,7 +3820,7 @@ BOOL CEEInfo::canInlineTypeCheckWithObjectVTable (CORINFO_CLASS_HANDLE clsHnd)
     if (VMClsHnd.IsTypeDesc())
     {
-        // We can't do this optimization for arrays because of the object methodtable is template methodtable
+        // We can't do this optimization for TypeDesc because methodtable could be shared
```
some TypeDescs still share method tables. I am not sure if such TypeDescs can occur on this code path, but perhaps it is safer to keep the check?

---
In reply to: [362353144](https://github.com/dotnet/runtime/pull/1201#discussion_r362353144) [](ancestors = 362353144)

jkotas replied:
The JIT or JITed code will never see the MethodTable behind these TypeDescs. I think it should be safe to delete this.

VSadov replied:
Will make `VMClsHnd.IsTypeDesc()` an assert

---
In reply to: [362378456](https://github.com/dotnet/runtime/pull/1201#discussion_r362378456) [](ancestors = 362378456)

jkotas replied:
I do not think you can assert this. The JIT may pass TypeDesc into this method in some cases.

My point was the JIT or JITed code never sees the underlying MethodTable that may be shared.

VSadov replied:
I see. It is possible to see a MethodDesc with shared MethodTable (ex: `typeof(int*) == o.GetType()`), but that is ok now. That is because optimized version compares literally _handle_ to the methodtable of the object and that produce valid result since none of remaining MethodDesc types can have heap instances.
We could as well just "optimize" the compare into `false` when the handle comparand is a TypeDesc, but it is such a corner case, so not worth the trouble.

---
In reply to: [362417850](https://github.com/dotnet/runtime/pull/1201#discussion_r362417850) [](ancestors = 362417850)

jkotas commented:
You should be also delete `ELEMENT_TYPE_NATIVE_ARRAY_TEMPLATE_ZAPSIG` and the code related to it. It should be no longer needed.

VSadov committed:
PR feedback:
Removed `CORINFO_HELP_NEWARR_1_R2R_DIRECT`
Folded a `cmp` in x86 array Address.
Removed` METHOD__BUFFER__BLOCKCOPY` from mscorlib.h
`GetMethodTableOfRootTypeParam`
`GetRootTypeParam`
Straightened `Object::GetTypeHandle`
`AsMethodTable` changes from PR feedback

VSadov committed:
Removed `ELEMENT_TYPE_NATIVE_ARRAY_TEMPLATE_ZAPSIG`

jkotas commented on src/coreclr/src/vm/prestub.cpp:
```diff
@@ -3241,8 +3241,7 @@ PCODE DynamicHelperFixup(TransitionBlock * pTransitionBlock, TADDR * pCell, DWOR
                 CorInfoHelpFunc helpFunc = CEEInfo::getNewArrHelperStatic(th);
-                ArrayTypeDesc *pArrayTypeDesc = th.AsArray();
-                MethodTable *pArrayMT = pArrayTypeDesc->GetTemplateMethodTable();
+                MethodTable *pArrayMT = th.GetMethodTable();
```
```suggestion
                MethodTable *pArrayMT = th.AsMethodTable();
```

jkotas commented on src/coreclr/src/vm/reflectioninvocation.cpp:
```diff
@@ -2408,7 +2402,7 @@ FCIMPL1(Object*, ReflectionSerialization::GetUninitializedObject, ReflectClassBa
+    if (type.IsTypeDesc() || type.IsArray())
         COMPlusThrow(kArgumentException, W("Argument_InvalidValue"));
 
     MethodTable *pMT = type.GetMethodTable();
```
```suggestion
    MethodTable *pMT = type.AsMethodTable();
```

VSadov committed:
`TypeDesc` is ok in `canInlineTypeCheckWithObjectVTable`

VSadov committed:
trivial  `AsMethodTable` changes

davidwrighton commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -3845,13 +3812,6 @@ BOOL CEEInfo::canInlineTypeCheckWithObjectVTable (CORINFO_CLASS_HANDLE clsHnd)
-        ret = FALSE;
-    }
-    else
     if (VMClsHnd == TypeHandle(g_pCanonMethodTableClass))
```
Logically this function should be able to simply return TRUE in all cases, but I'm not sure if the jit can tolerate that in the case of a shared generic.. If it can't, please file an issue against the jit about that.

(The interesting case is)

if (typeof(T) == obj.GetType()) {...}

VSadov replied:
`typeof(T) == obj.GetType()`  does not get here. 
`gtFoldTypeCompare` requires that at least one comparand is effectively a constant (`TPK_Handle`) and `typeof(T)` for a generic typeparam is not a constant.

In addition to that, it is actually classified as `TPK_Unknown`. No idea why, I did not dig further.
I would expect `TPK_GetType`.

I agree this is now optimizable. - as long as shared generics are used only for reference types.  (since arrays are no longer special).

I will log a bug for the JIT.

VSadov replied:
Logged [https://github.com/dotnet/runtime/issues/1258](https://github.com/dotnet/runtime/issues/1258)


---
In reply to: [362655670](https://github.com/dotnet/runtime/pull/1201#discussion_r362655670) [](ancestors = 362655670)

davidwrighton commented on src/coreclr/src/vm/object.cpp:
```diff
@@ -210,20 +194,16 @@ TypeHandle Object::GetGCSafeTypeHandleIfPossible() const
     MethodTable * pMTToCheck = pMT;
     if (pMTToCheck->IsArray())
     {
```
This entire special case is no longer necessary.

VSadov replied:
What makes it different? Is that because TypeDesc was not preventing element type unload and MethodTable prevents that?

If that is the case, I am guessing the whole `GetGCSafeTypeHandleIfPossible` can be removed.

davidwrighton replied:
Its different as ArrayTypeDesc::GetLoaderModule could walk through partially unloaded data structures. However, MethodTable::GetLoaderModule doesn't have that problem.

We still need GetGCSafeTypeHandleIfPossible to handle the collectiable assemblies cases. 

---
In reply to: [362658395](https://github.com/dotnet/runtime/pull/1201#discussion_r362658395) [](ancestors = 362658395)

davidwrighton commented on src/coreclr/src/vm/olevariant.cpp:
```diff
@@ -300,7 +300,7 @@ VARTYPE OleVariant::GetVarTypeForTypeHandle(TypeHandle type)
 
     // Handle objects.
-    if (!type.IsTypeDesc())
+    if (!type.IsArray())
```
This should also exclude TypeDescs as before.

VSadov replied:
Right. Good catch!

---
In reply to: [362655649](https://github.com/dotnet/runtime/pull/1201#discussion_r362655649) [](ancestors = 362655649)

davidwrighton commented on src/coreclr/src/vm/typehandle.cpp:
```diff
@@ -468,22 +452,18 @@ BOOL TypeHandle::IsBlittable() const
 
+    if (IsArray())
+    {
+        // Single dimentional array's of blittable types are also blittable.
```
@jkoritzinsky Is this the semantic? Should 1 dimensional arrays with non-zero lower bounds be considered blittable?

VSadov replied:
Supporting NZLB arrays does seem a bit strange here.

I will consider this resolved though, since there is no change in behavior. 

---
In reply to: [362658073](https://github.com/dotnet/runtime/pull/1201#discussion_r362658073) [](ancestors = 362658073)

VSadov committed:
`GetVarTypeForTypeHandle` should throw `unsupported` on TypeDescs

davidwrighton commented on src/coreclr/src/vm/methodtable.h:
```diff
@@ -1973,9 +1973,9 @@ class MethodTable
+    BOOL ArrayIsInstanceOf(MethodTable* pTargetMT, TypeHandlePairList* pVisited);
 
     BOOL CanCastByVarianceToInterfaceOrDelegate(MethodTable* pTargetMT, TypeHandlePairList* pVisited);
 
```
Looking through GetGuid, it appears there are paths which reach IsLegalNonArrayWinRTType without an array check. I'm not sure if its possible to actually get there, but it feels like Guid management and this change is a risky combination. (At the very least, I feel there need to be more asserts that invalid paths aren't taken.)

VSadov replied:
I looked through callers to `IsLegalNonArrayWinRTType` and it seems seems to be called only after being sure the type is not an array. 
Some subset of arrays is actually supported in WinRT scenarios and, if relevant, arrays are handled before calling `IsLegalNonArrayWinRTType`.

In the majority of cases the type is known to be an interface or a delegate though.

VSadov replied:
`ReflectionInvocation::GetGUID` does have a bug. Previously it would produce 0000...00 guid for all typedescs. Now it returns a guid for arrays. 

Interestingly, there is also `FCIMPL2(void, MarshalNative::DoGenerateGuidForType` which looks like supposed to work with arrays and produce guids, but I m not sure if there are any callers.

jkotas replied:
> `MarshalNative::DoGenerateGuidForType`

This is dead code. You can delete it while you are on it.

VSadov committed:
not digging for the root element type in `GetGCSafeTypeHandleIfPossible`

jkotas commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -3861,6 +3821,9 @@ BOOL CEEInfo::canInlineTypeCheckWithObjectVTable (CORINFO_CLASS_HANDLE clsHnd)
     else
     {
         // It is safe to perform this optimization
+        // NOTE: clsHnd could be a MethodDesc with shared MethodTable (ex: typeof(int*) == o.GetType()), but that is still optimizable.
```
```suggestion
        // NOTE: clsHnd could be a TypeDesc with shared MethodTable (ex: typeof(int*) == o.GetType()), but that is still optimizable.
```

jkotas commented on src/coreclr/src/vm/object.cpp:
```diff
@@ -208,24 +192,6 @@ TypeHandle Object::GetGCSafeTypeHandleIfPossible() const
     //         Object[], in which case its MT is in mscorlib and thus doesn't unload.
 
     MethodTable * pMTToCheck = pMT;
```
`pMTToCheck` can be deleted and `pMT` used instead.

jkotas commented on src/coreclr/src/vm/olevariant.cpp:
```diff
@@ -400,14 +403,8 @@ VARTYPE OleVariant::GetVarTypeForTypeHandle(TypeHandle type)
-    }
-
-    return VT_ARRAY;
+    // Non interop compatible type.
```
Nit: It may be nice to move this up to make the code more linear:

```
if (type.IsTypeDesc())
    COMPlusThrow(kArgumentException, IDS_EE_COM_UNSUPPORTED_SIG);
```

VSadov committed:
Last PR feedback

jkotas commented on src/coreclr/src/vm/methodtable.h:
```diff
@@ -468,15 +468,16 @@ struct MethodTableWriteableData
-                       MethodTableWriteableData::enum_flag_Unrestored |
-                       MethodTableWriteableData::enum_flag_IsNotFullyLoaded |
-                       MethodTableWriteableData::enum_flag_HasApproxParent);
+        SetIsRestoredForBuildMethodTable();
```
What motivated this change?

VSadov replied:
To match the initial load level of ArrayTypeDesc.
Otherwise the array MT is always fully loaded, even if when the element type may be not fully loaded.
I did not see such cases, but it seems possible.

VSadov replied:
This will also force some checks that are done when moving to fully loaded state - like that the type is not infinitely recursive.

---
In reply to: [362986502](https://github.com/dotnet/runtime/pull/1201#discussion_r362986502) [](ancestors = 362986502)

VSadov committed:
Fix `TypeInfoTests.Guid()`

VSadov commented:
@jkotas @davidwrighton  - any more comments on this?

jkotas reviewed (approved):
LGTM. Thank you for making this happen!

mjsabby commented:
After this change, will (typeHandle & 0x2) ever be true?

jkotas commented:
Yes, for pointer, byref, generic variables, etc.

VSadov commented:
Thanks!!!

VSadov commented:
@mjsabby  the nice part is that any standalone/boxed/heap instance can assume that its MethodTable is also its TypeHandle and also its unique and only type identity.

There are many scenarios that get cheaper or simpler if this holds.
The equivalency was mostly the case already, except for arrays.

MichalStrehovsky commented on src/coreclr/src/inc/corinfo.h:
```diff
@@ -407,7 +407,6 @@ enum CorInfoHelpFunc
     CORINFO_HELP_NEW_MDARR,         // multi-dim array helper (with or without lower bounds - dimensions passed in as vararg)
     CORINFO_HELP_NEW_MDARR_NONVARARG,// multi-dim array helper (with or without lower bounds - dimensions passed in as unmanaged array)
     CORINFO_HELP_NEWARR_1_DIRECT,   // helper for any one dimensional array creation
-    CORINFO_HELP_NEWARR_1_R2R_DIRECT, // wrapper for R2R direct call, which extracts method table from ArrayTypeDesc
```
Since this updated corinfo.h, this should have also revved the JitEEVersionIdentifier: see https://github.com/dotnet/runtime/blob/master/docs/project/updating-jitinterface.md

VSadov replied:
A PR to update the guid: https://github.com/dotnet/runtime/pull/1330

mjsabby commented:
> @mjsabby the nice part is that any standalone/boxed/heap instance can assume that its MethodTable is also its TypeHandle and is also its unique and only type identity.
> 
> There are many scenarios that get cheaper or simpler if this holds.
> The equivalency was mostly the case already, except for arrays.

@VSadov Yup I asked precisely because I only care about objects that may be in a GC heap, and I had written some code that made this assumption. No longer true for 5.0!

cshung commented on src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/R2RSignature.cs:
```diff
@@ -1054,10 +1054,6 @@ private void ParseType(StringBuilder builder)
                     builder.Append("var_zapsig");
                     break;
 
-                case CorElementType.ELEMENT_TYPE_NATIVE_ARRAY_TEMPLATE_ZAPSIG:
```
This code path is used to parse signatures in ready-to-run images.

As far as I understand, ready-to-run images are supposed to be forward compatible. Old crossgen compiled ready-to-run images should run on a newer version of the runtime. Changing the signature format seems to break things.

@VSadov @jkotas 

At the very least, R2RDump should still be able to dump the content of the older version of ready to run images, thoughts?

jkotas commented:
I do not think that `ELEMENT_TYPE_NATIVE_ARRAY_TEMPLATE_ZAPSIG` was ever emitted in R2R images. Notice that the place that emitted it had an explicit condition around it: https://github.com/dotnet/runtime/pull/1201/files#diff-bf3aa4d205b05f175ed65d245135fc73L3327

cshung commented:
> I do not think that `ELEMENT_TYPE_NATIVE_ARRAY_TEMPLATE_ZAPSIG` was ever emitted in R2R images. 

I have a copy of a `System.Private.CoreLib.dll` built using `crossgen`, `x64`, `debug` that contains it.
It was built locally from commit `d099bc93f9abd814dfae0e9dc9999c26508faf7e`.

The elimination of `ELEMENT_TYPE_NATIVE_ARRAY_TEMPLATE_ZAPSIG` is causing a diff in R2RDump output before and after the change.

jkotas commented:
Was it for in the signature for `byte[]` in `Buffer.BlockCopy`, or something else?

System.Private.CoreLib.dll has number of special casing because it versions with the runtime. Do you see it anywhere else outside of System.Private.CoreLib.dll ?

cshung commented:
> Was it for in the signature for `byte[]` in `Buffer.BlockCopy`, or something else?
> 
> System.Private.CoreLib.dll has number of special casing because it versions with the runtime. Do you see it anywhere else outside of System.Private.CoreLib.dll ?

Yes, that the only one I have discovered so far.

jkotas commented:
The one in `Buffer.BlockCopy` was introduced just recently. It never shipped. I do not think it is anything to worry about.
